core/shared/src/main/scala/zio/ZIO.scala

```diff
@@ -1539,11 +1539,14 @@ sealed trait ZIO[-R, +E, +A]
    */
   final def repeatN(n: => Int)(implicit trace: Trace): ZIO[R, E, A] =
     ZIO.suspendSucceed {
+      var n0 = n
 
-      def loop(n: Int): ZIO[R, E, A] =
-        self.flatMap(a => if (n <= 0) Exit.succeed(a) else ZIO.yieldNow *> loop(n - 1))
-
-      loop(n)
+      if (n0 <= 0) self
+      if (n0 == 1) {
+        self *> self
+      } else {
+        ZIO.whileLoop(n0 != 0)(self)(_ => n0 -= 1) *> self
+      }
```

Comment:
While I was running the benchmarks, I realised the performance of `ZIO#repeatN` was quite bad. This improves it massively. What do we think of it?

Reply:
👍

Reply:
PS: Opened #9082 to track this separately

Reply:
Sounds good but I would separate it from this PR.

Reply:
Sounds good, I'll revert this change for now